src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -113,7 +114,15 @@ public void Write(SourceAwareJsonTextWriter writer)
 
             this.EmitMetadata(emitter, program.Metadata);
 
-            this.EmitTypeDefinitionsIfPresent(emitter, program.Types.Concat(ImportClosureInfo.ImportedTypesInClosure));
+            var programTypes = program.Types.Concat(ImportClosureInfo.ImportedTypesInClosure);
+
+            if (programTypes.Any())
+            {
+                TypeAliasLookupVisitor = new ExpressionTypeAliasLookupVisitor();
+                TypeAliasLookupVisitor.Visit(program);
+
+                this.EmitTypeDefinitionsIfPresent(emitter, programTypes);
```

Comment:
I would pass `TypeAliasLookupVisitor` to `EmitTypeDefinitionsIfPresent` rather than binding it to the class. Conditionally binding it like this requires any usage in that method to either do a null check or use `!`, which is a bit of a footgun since it'll be on PR reviewers rather than the compiler to figure out when the `!` is actually unsafe.

Reply:
https://github.com/Azure/bicep/pull/11222/commits/b889fea2e5d0e05eff348307313020b7cf24009b
I converted this to a dictionary per the other comment. However, I'm still leaning on keeping this an instance variable for now as an argument becomes very messy as it has to go through `TypePropertiesForTypeExpression`. I moved the initializer further up to before the top-level template object is written.